flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/expressions/ExpressionExtractionUtils.java

```diff
@@ -0,0 +1,211 @@
+			.collect(Collectors.toList());
+	}
+
+	private static class AggregationAndPropertiesReplacer extends DefaultExpressionVisitor<Expression> {
+
+		private final Map<Expression, String> aggregates;
+		private final Map<Expression, String> properties;
+
+		private AggregationAndPropertiesReplacer(
+				Map<Expression, String> aggregates,
+				Map<Expression, String> properties) {
+			this.aggregates = aggregates;
+			this.properties = properties;
+		}
+
+		@Override
+		public Expression visitLookupCall(LookupCallExpression unresolvedCall) {
+			throw new IllegalStateException("All calls should be resolved by now. Got: " + unresolvedCall);
+		}
+
+		@Override
+		public Expression visitCall(CallExpression call) {
+			if (aggregates.get(call) != null) {
+				return unresolvedFieldRef(aggregates.get(call));
+			} else if (properties.get(call) != null) {
+				return unresolvedFieldRef(properties.get(call));
+			}
+
+			List<Expression> args = call.getChildren()
+				.stream()
+				.map(c -> c.accept(this))
+				.collect(Collectors.toList());
+			return new CallExpression(call.getFunctionDefinition(), args);
+		}
+
+		@Override
+		protected Expression defaultMethod(Expression expression) {
+			return expression;
+		}
+	}
+
+	/**
+	 * Extract all field references from the given expressions.
+	 *
+	 * @param expressions a list of expressions to extract
+	 * @return a list of field references extracted from the given expressions
+	 */
+	public static List<Expression> extractFieldReferences(List<Expression> expressions) {
+		FieldReferenceExtractor referenceExtractor = new FieldReferenceExtractor();
+		return expressions.stream()
+			.flatMap(expr -> expr.accept(referenceExtractor).stream())
+			.distinct()
+			.collect(Collectors.toList());
+	}
+
+	private static class FieldReferenceExtractor extends DefaultExpressionVisitor<List<Expression>> {
+
+		@Override
+		public List<Expression> visitCall(CallExpression call) {
+			FunctionDefinition functionDefinition = call.getFunctionDefinition();
+			if (WINDOW_PROPERTIES.contains(functionDefinition)) {
+				return Collections.emptyList();
+			} else {
+				return call.getChildren()
+					.stream()
+					.flatMap(c -> c.accept(this).stream())
+					.distinct()
+					.collect(Collectors.toList());
+			}
+		}
+
+		@Override
+		public List<Expression> visitLookupCall(LookupCallExpression unresolvedCall) {
+			throw new IllegalStateException("All calls should be resolved by now. Got: " + unresolvedCall);
+		}
+
+		@Override
+		public List<Expression> visitFieldReference(FieldReferenceExpression fieldReference) {
+			return Collections.singletonList(fieldReference);
+		}
+
+		@Override
+		public List<Expression> visitUnresolvedField(UnresolvedFieldReferenceExpression fieldReference) {
+			return Collections.singletonList(fieldReference);
+		}
+
+		@Override
+		protected List<Expression> defaultMethod(Expression expression) {
+			return Collections.emptyList();
+		}
+	}
+
+	private ExpressionExtractionUtils() {
```

Comment:
Move this up? I was just about to write a comment that it is missing.

Reply:
Personally I prefer having the private ctor of a utility class at the end. IMHO it is a boiler plate code that distracts from the essence of the class at the beginning.

From a quick scan of *Util* classes we usually have the ctor at the end.

Reply:
I wasn't aware of that. Maybe this is personal taste. IMHO it helps to see immediately that this class is not instantiable at the very beginning. member variables, constructors, methods should always be the correct sequence.

Reply:
Agree if there are variables, multiple ctors etc.

This is a special case though. It will never have member variables, no non-static functions and always a single private empty ctor. For me Utils name is enough to say the class should not be instantiable. Whenever I see Utils in the name of the class. The first place where I look the private ctor is at the end.